lib/linguist/heuristics.rb

```diff
@@ -13,11 +13,14 @@ class Heuristics
     #   ])
     #
     # Returns an Array of languages, or empty if none matched or were inconclusive.
-    def self.call(blob, languages)
+    def self.call(blob, candidates)
       data = blob.data
 
       @heuristics.each do |heuristic|
-        return Array(heuristic.call(data)) if heuristic.matches?(languages)
+        if heuristic.matches?(blob.name)
+          languages = Array(heuristic.call(data))
+          return languages if languages.any? || languages.all? { |l| candidates.include?(l) }
```

Comment:
`languages.any? || languages.all? { |l| candidates.include?(l) }` should always be true.
Maybe we should throw a warning when it's not...?